src/perl/Permabit/BlockDevice/LVM/Thin.pm

```diff
@@ -15,12 +15,40 @@ use base qw(Permabit::BlockDevice::LVM);
 
 my $log = Log::Log4perl->get_logger(__PACKAGE__);
 
+########################################################################
+# @inherit
+##
+sub checkStorageDevice {
+  my ($self) = assertNumArgs(1, @_);
+  # This will create the bottom device if needed.
+  $self->SUPER::checkStorageDevice();
+  # If our storage device is a thin pool (this includes vdo thin pools),
+  # don't do anything, just use it.
+  if ($self->{storageDevice}->isa("Permabit::BlockDevice::LVM::ThinPool")) {
+    $self->{volumeGroup} = $self->{storageDevice}->{volumeGroup};
+    return;
+  }
+  # If we don't have a thin pool already, create a generic one.
+  $self->{storageDevice} = $self->{stack}->create("thinpool")
+}
+
+########################################################################
+# @inherit
+##
+sub configure {
+  my ($self, $arguments) = assertNumArgs(2, @_);
+  $self->SUPER::configure($arguments);
+  $self->{lvmSize} = $self->{storageDevice}->getSize();
+  $self->{lvmSize} = $self->{volumeGroup}->alignToExtent($self->{lvmSize});
```

Comment:
> This spelling seems odd to me. The SUPER method also tries to set lvmSize so you're actually doing it twice by copying these lines. Also, since we have the optional $arguments hash, couldn't you just add lvmSize to the hash of arguments, especially since you're setting it unilaterally? That might be cleaner.

True. The wording was bad and not consistent with the code. I've reworked both to make it more understandable.

Reply:
The problem here is that the thinpool (if it doesn't already exist) is created in the configure method of thin.pm in the BlockDevice::configure method when it calls checkStorageDevice. So we are waiting till the super::configure is done before we can get the size of the thin pool which is what we want to use as the size for the thin volumes. I'm not sure what the best solution would be here. We could assert the thinpool is already created before and don't auto create it, but I put that there so that all our old tests which just use thin as the device type would just "work" with the new setup.

Reply:
I don't know why I didn't mention this before. Couldn't you add a parameter that basically says "Use the full size of the pool"? That's really what you want here, right? I get that you don't have a size you can pass in, but can't we just tell the superclass to Do The Right Thing?

Reply:
I think I figured out a way to do this without the need for a configure override. I've put the setting of lvmsize in checkStorageDevice so that its set to the pool device size (not the volume group size as LVM.pm does).